Upload pretrained models

AlibiZhenis opened the pull request:
### Description
Added a feature of uploading pretrained models. Closes #106 .
 
### Check List
- [x] New functionality includes testing.
  - [ ] All tests pass
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

AlibiZhenis commented on opensearch_py_ml/ml_commons/ml_commons_client.py:
```diff
@@ -63,6 +63,71 @@ def upload_model(
             model_path, model_config_path, isVerbose
         )
 
+    def upload_pretrained_model(self, name: str, version: str, model_format):
```
So, if I'm not mistaken, to upload a pretrained model, we have to:

1. Send model info to /models/_upload
2. Then load the model using load model api, after which ML Commons downloads the model chunks from Hugging Face by itself.

If I do the above 2 steps, I can successfully use the model for inference in dev tools.

Are you saying that that I shouldn't do the second part? Because without it, if I try to do inference, I get the "model not loaded" error. 

I would appreciate it if you clarified this for me.

AlibiZhenis replied:
I think it's a great idea. The argument was added.

Also, about sending model info to upload api, that's what _create_model_metadoc method does anyway. I just misnamed the method.

Both of these issues were fixed in the last commit.

dhrubo-os commented:
Thanks for the PR. It looks good. 

Integration test is failing. I believe this is because right now our integration test is based on [opensearch 2.5.0](https://github.com/opensearch-project/opensearch-py-ml/blob/main/.github/workflows/integration.yml#L15)

But this new feature is in `opensearch 2.6.0`. So we should update that version to pass the integration test.

Lint check is also failing. [Check here](https://github.com/opensearch-project/opensearch-py-ml/blob/main/DEVELOPER_GUIDE.md#to-test-code-formatting-and-linting-issue-we-can-run)

AlibiZhenis commented:
So what should I do next? I have opensearch-py 2.2.0 installed, which pip says is the latest version.

dhrubo-os commented:
> So what should I do next? I have opensearch-py 2.2.0 installed, which pip says is the latest version.



> 

Can you please change to `2.6.0` [here])https://github.com/opensearch-project/opensearch-py-ml/blob/main/.github/workflows/integration.yml#L15) ?

dhrubo-os commented:
You can also run the whole test in your end like this: `./.ci/run-tests "opensearch" "true" 2.6.0` to check

AlibiZhenis commented:
It seems the test was failing because model unload was timing out. Once I increased the threshold to 60s, it's now passing on my side.

dhrubo-os commented on opensearch_py_ml/ml_commons/ml_commons_client.py:
```diff
@@ -63,6 +63,76 @@ def upload_model(
+        """
+        This method uploads a pretrained model into opensearch cluster using ml-common plugin's api.
+        First, this method creates a model id to store model info
+        and then loads the model from Hugging Face if load_model is True.
```
then loads the model in memory. We don't load model from huggingFace. Model is downloaded in their index and then it will be loaded into memory using DJL.

dhrubo-os commented on opensearch_py_ml/ml_commons/ml_commons_client.py:
```diff
@@ -63,6 +63,76 @@ def upload_model(
         )
 
+    def upload_pretrained_model(
+        self, name: str, version: str, model_format: str, load_model: bool = True
```
how about `model_name`, `model_version`?

dhrubo-os commented on opensearch_py_ml/ml_commons/ml_commons_client.py:
```diff
@@ -63,6 +63,76 @@ def upload_model(
+        :type name: string
+        :param version: Version of the pretrained model
+        :type version: string
+        :param model_format: Currently only supports "TORCH_SCRIPT"
```
We also support `ONNX`. If you see this : https://opensearch.org/docs/latest/ml-commons-plugin/pretrained-models/, we also have urls for `ONNX` artifacts.

dhrubo-os commented on opensearch_py_ml/ml_commons/ml_commons_client.py:
```diff
@@ -63,6 +63,76 @@ def upload_model(
+        :type version: string
+        :param model_format: Currently only supports "TORCH_SCRIPT"
+        :type model_format: string
+        :param load_model: Whether to load the model chunks from Hugging Face
```
whether to load the model in memory using uploaded model chunks.

dhrubo-os commented on opensearch_py_ml/ml_commons/ml_commons_client.py:
```diff
@@ -63,6 +63,76 @@ def upload_model(
+        This method sends the pretrained model info to ML Commons' upload api
+
+        :param model_meta_json: a dictionary object with model configurations
+        :type model_meta_json: string
```
dict instead of string

dhrubo-os commented on opensearch_py_ml/ml_commons/ml_commons_client.py:
```diff
@@ -63,6 +63,76 @@ def upload_model(
+        :param model_meta_json: a dictionary object with model configurations
+        :type model_meta_json: string
+        :return: returns a unique id of the model
+        :rtype: int
```
string instead of int

dhrubo-os commented:
When you push a new commit with all these changes can you also run the workflow. Please let me know if you don't have permission to run workflow?

AlibiZhenis commented:
Sorry for all the mistakes. I'll try to be more attentive from now on.

It also seems that I don't have permission to run workflows.

dhrubo-os commented on tests/ml_commons/test_ml_commons_client.py:
```diff
@@ -70,6 +74,47 @@ def test_init():
+            ml_model_status = ml_client.get_model_info(model_id)
+            print("ml_model_status", ml_model_status)
+            assert ml_model_status.get("model_state") == "UNLOADED"
+        except:
```
Lint is failing for bare except, you need to add to skip linting error `# noqa: E722`

dhrubo-os commented on tests/ml_commons/test_ml_commons_client.py:
```diff
@@ -70,6 +74,47 @@ def test_init():
+        raised = False
+        try:
+            ml_client.unload_model(model_id)
+            time.sleep(60)
```
Integration is still failing. May be we can have a loop to check back in every 15 seconds. if we get the result, that's great otherwise will wait for another 15 seconds. 

But we can't run this forever. May be the hard limit can be 5 minutes. IF total duration passes more than 5 minutes then fail the test.

dhrubo-os commented:
> > So what should I do next? I have opensearch-py 2.2.0 installed, which pip says is the latest version.
> 
> Can you please change to `2.6.0` [here](https://github.com/opensearch-project/opensearch-py-ml/blob/main/.github/workflows/integration.yml#L15) ?

I saw you changed the version, but now i don't see the changed code. Where is it? Integration test is failing....

dhrubo-os commented on .github/workflows/integration.yml:
```diff
@@ -12,7 +12,7 @@ jobs:
         secured: ["true"]
         entry:
-          - { opensearch_version: 2.6.0 }
+          - { opensearch_version: 2.5.0 }
```
Why did you change this back?

AlibiZhenis replied:
The tests are passing in my local environment, but not the workflow. So I wanted to see if the problem was with the version. Turns out it's not, I will change it back.

I am suspecting the problem is with the circuit breaker error. I am investigating.

AlibiZhenis commented:
It seems to be consistently passing now once I changed the version back to 2.6.0. I believe the reason is in my "minor code and test changes", where I changed the upload_pretrained_model method.

Now the method finishes once the model_state becomes loaded, instead of when the load model task gets completed.
